File: packages/cloud-adapters/secrets/baremetal.go
package secrets

import (
	"context"
	"encoding/base64"
	"fmt"

	vault "github.com/hashicorp/vault/api"
)

// BaremetalAdapter is the Vault-Transit-backed secrets adapter. The
// Transit Secrets Engine is per-tenant by convention — keys are named
// per-tenant (e.g. `tenant-acme-pii`), and crypto-shred = delete-key
// (Discussion #76). Mount path is fixed at construction; multi-mount
// use cases get one adapter per mount.
type BaremetalAdapter struct {
	client *vault.Client
	mount  string
}

// NewBaremetalAdapter constructs a Vault-Transit-backed adapter. The
// caller passes a *vault.Client already configured with endpoint +
// token (typically via vault.DefaultConfig() + client.SetToken()).
// The Transit engine must already be mounted at `mount` (operator
// runbook responsibility; not part of the adapter).
func NewBaremetalAdapter(client *vault.Client, mount string) *BaremetalAdapter {
	return &BaremetalAdapter{client: client, mount: mount}
}

func (a *BaremetalAdapter) MintKey(ctx context.Context, name string) error {
	path := fmt.Sprintf("%s/keys/%s", a.mount, name)
	if _, err := a.client.Logical().WriteWithContext(ctx, path, map[string]any{
		"type": "aes256-gcm96",
	}); err != nil {
		return fmt.Errorf("baremetal secrets MintKey %q: %w", name, err)
	}
	return nil
}

func (a *BaremetalAdapter) Encrypt(ctx context.Context, name string, plaintext []byte) (string, error) {
	path := fmt.Sprintf("%s/encrypt/%s", a.mount, name)
	res, err := a.client.Logical().WriteWithContext(ctx, path, map[string]any{
		"plaintext": base64.StdEncoding.EncodeToString(plaintext),
	})
	if err != nil {
		return "", fmt.Errorf("baremetal secrets Encrypt %q: %w", name, err)
	}
	if res == nil || res.Data == nil {
		return "", fmt.Errorf("baremetal secrets Encrypt %q: empty response", name)
	}
	ct, ok := res.Data["ciphertext"].(string)
	if !ok {
		return "", fmt.Errorf("baremetal secrets Encrypt %q: missing ciphertext", name)
	}
	return ct, nil
}

func (a *BaremetalAdapter) Decrypt(ctx context.Context, name string, ciphertext string) ([]byte, error) {
	path := fmt.Sprintf("%s/decrypt/%s", a.mount, name)
	res, err := a.client.Logical().WriteWithContext(ctx, path, map[string]any{
		"ciphertext": ciphertext,
	})
	if err != nil {
		return nil, fmt.Errorf("baremetal secrets Decrypt %q: %w", name, err)
	}
	if res == nil || res.Data == nil {
		return nil, fmt.Errorf("baremetal secrets Decrypt %q: empty response", name)
	}
	b64, ok := res.Data["plaintext"].(string)
	if !ok {
		return nil, fmt.Errorf("baremetal secrets Decrypt %q: missing plaintext", name)
	}
	pt, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return nil, fmt.Errorf("baremetal secrets Decrypt %q: invalid base64: %w", name, err)
	}
	return pt, nil
}

func (a *BaremetalAdapter) RotateKey(ctx context.Context, name string) error {
	path := fmt.Sprintf("%s/keys/%s/rotate", a.mount, name)
	if _, err := a.client.Logical().WriteWithContext(ctx, path, nil); err != nil {
		return fmt.Errorf("baremetal secrets RotateKey %q: %w", name, err)
	}
	return nil
}

func (a *BaremetalAdapter) DeleteKey(ctx context.Context, name string) error {
	// Vault refuses to delete a transit key unless deletion is enabled
	// on the key itself (`allow_plaintext_backup` / `deletion_allowed`).
	// Per the crypto-shred design (Discussion #76), the adapter flips
	// `deletion_allowed=true` first, then deletes.
	cfgPath := fmt.Sprintf("%s/keys/%s/config", a.mount, name)
	if _, err := a.client.Logical().WriteWithContext(ctx, cfgPath, map[string]any{
		"deletion_allowed": true,
	}); err != nil {
		return fmt.Errorf("baremetal secrets DeleteKey %q: enable deletion: %w", name, err)
	}
	path := fmt.Sprintf("%s/keys/%s", a.mount, name)
	if _, err := a.client.Logical().DeleteWithContext(ctx, path); err != nil {
		return fmt.Errorf("baremetal secrets DeleteKey %q: %w", name, err)
	}
	return nil
}

// Compile-time interface conformance check.
var _ Adapter = (*BaremetalAdapter)(nil)

File: packages/cloud-adapters/secrets/baremetal.ts
import type { SecretsAdapter } from "./interface";

// Configuration the operator passes to the factory. `fetch` is
// injectable so tests can stub the HTTP call without touching the
// global; production code passes the default global `fetch`.
export interface VaultConfig {
  endpoint: string; // e.g. "https://vault.svc.cluster.local:8200"
  token: string;
  mount: string;   // Transit engine mount path, e.g. "transit"
  fetch?: typeof fetch;
}

// Vault wraps every response in `{ request_id, data }`. The Transit
// engine returns `ciphertext` / `plaintext` (base64) inside `data`.
interface VaultResponse {
  data?: {
    ciphertext?: string;
    plaintext?: string;
  };
  errors?: string[];
}

const utf8 = new TextDecoder();

const toBase64 = (b: Uint8Array): string =>
  Buffer.from(b.buffer, b.byteOffset, b.byteLength).toString("base64");

const fromBase64 = (s: string): Uint8Array =>
  Uint8Array.from(Buffer.from(s, "base64"));

export function newBaremetalAdapter(cfg: VaultConfig): SecretsAdapter {
  const f = cfg.fetch ?? fetch;
  const headers = { "X-Vault-Token": cfg.token, "Content-Type": "application/json" };
  const base = `${cfg.endpoint}/v1/${cfg.mount}`;

  // call wraps the Vault request envelope + error surfacing once.
  const call = async (
    path: string,
    method: "POST" | "DELETE",
    body: object | null,
  ): Promise<VaultResponse> => {
    const res = await f(`${base}/${path}`, {
      method,
      headers,
      body: body == null ? undefined : JSON.stringify(body),
    });
    if (!res.ok) {
      // Vault returns errors as JSON; fall through to text() if it isn't.
      const raw = await res.text();
      throw new Error(`vault transit ${method} ${path} → ${res.status}: ${raw}`);
    }
    // 204 No Content (e.g. DeleteKey) has no body to parse.
    if (res.status === 204) return {};
    return (await res.json()) as VaultResponse;
  };

  return {
    async mintKey(name: string): Promise<void> {
      await call(`keys/${encodeURIComponent(name)}`, "POST", { type: "aes256-gcm96" });
    },
    async encrypt(name: string, plaintext: Uint8Array): Promise<string> {
      const r = await call(`encrypt/${encodeURIComponent(name)}`, "POST", {
        plaintext: toBase64(plaintext),
      });
      const ct = r.data?.ciphertext;
      if (!ct) throw new Error(`vault transit encrypt ${name}: missing ciphertext`);
      return ct;
    },
    async decrypt(name: string, ciphertext: string): Promise<Uint8Array> {
      const r = await call(`decrypt/${encodeURIComponent(name)}`, "POST", { ciphertext });
      const pt = r.data?.plaintext;
      if (pt == null) throw new Error(`vault transit decrypt ${name}: missing plaintext`);
      return fromBase64(pt);
    },
    async rotateKey(name: string): Promise<void> {
      await call(`keys/${encodeURIComponent(name)}/rotate`, "POST", {});
    },
    async deleteKey(name: string): Promise<void> {
      // Per the crypto-shred design: flip deletion_allowed first,
      // then DELETE. Vault refuses key delete otherwise.
      await call(`keys/${encodeURIComponent(name)}/config`, "POST", {
        deletion_allowed: true,
      });
      await call(`keys/${encodeURIComponent(name)}`, "DELETE", null);
    },
  };
}

// Backwards-compat stub from the #80 scaffolding — throws on call;
// operators wire the real adapter via newBaremetalAdapter(cfg) instead.
const stubMsg = "cloud-adapters/secrets/baremetal: configure via newBaremetalAdapter(cfg)";
export const baremetal: SecretsAdapter = {
  mintKey:   async () => { throw new Error(stubMsg); },
  encrypt:   async () => { throw new Error(stubMsg); },
  decrypt:   async () => { throw new Error(stubMsg); },
  rotateKey: async () => { throw new Error(stubMsg); },
  deleteKey: async () => { throw new Error(stubMsg); },
};

// Suppress unused-import warning when the utf8 decoder isn't needed
// at runtime (kept available for callers that want to lift bytes to
// strings without going back through Buffer themselves).
export const _internal = { utf8 };
